### ISSUE TITLE
create fetchpypi function that generates name based urls for the prefetch sheme

### ISSUE THREAD

**Issue by RonnyPfannschmidt** [NixOS/nixpkgs#22256]
followup to #21637

modern pypi supports name based dispatch of different distribution types,

based on the distribution name, the distribution type (source, py2, py3, py2.3) and some extra details, 
pypi can be asked for the real file

however the current mirror url for pypi hardcodes the type source

```
fetchpypi = {name, version, type ? "source"}: ...
```

since nixos will always require building binary wheels, it seems sensible to have functions fitting the python

```
python.fetchWheel # choosing py2/py3 type based on python version
python.fetchUniversalWheel # for the py2.py3 type
python.fetchSdist # chooses source tarball
python.fetchSdistZip # for stranger uploads
```

**Comment by RonnyPfannschmidt** [comment]
pinging @FRidh

**Comment by FRidh** [comment]
How can we retrieve a wheel given name a version? As far as I know you have to know the hash to build up an url.

**Comment by RonnyPfannschmidt** [comment]
@FRidh relapce the "source" segment in the pypi.io url with the wheel type (py2, py3, py2.py3)

im working on a pr

**Comment by FRidh** [comment]
can you show a couple of sample urls

**Comment by RonnyPfannschmidt** [comment]
https://files.pythonhosted.org/packages/py2.py3/m/matlab_kernel/matlab_kernel-0.9.5-py2.py3-none-any.whl

**Comment by RonnyPfannschmidt** [comment]
https://files.pythonhosted.org/packages/py2.py3/s/setuptools/setuptools-34.0.0-py2.py3-none-any.whl

**Comment by FRidh** [comment]
Thanks for your examples. I don't think we should have that many different fetchers. I've merged 695ff0dc09b304f2e0575081c52189893c2f38bc. While `fetchtarball` and `fetchwheel` are public now we might want to change that.
